### ISSUE TITLE
How to use it in azure function?

### ISSUE THREAD

**Issue by taipignas** [EntityGraphQL/EntityGraphQL#352]
I read the documentation and it says that i can use this outside ASP.NET. I am trying to implement it in azure functions, but the usage of 'HttpContext.RequestServices' looks a bit strange to me and i dont know how to get around it. Is it possible to use this with isolated out of process azure functions? Also maybe AddGraphQLSchema shouldnt be in AspNet part of the library? Because its very useful for dependency injection and has nothing to do with ASP.NET itself, no?

**Comment by lukemurray** [comment]
https://entitygraphql.github.io/docs/getting-started#custom-controller--manual-execution this shows how to do it with a ASP.NET controller but you should be able to adapt that to something else.

**Comment by taipignas** [comment]
yes thats the example i tried to adapt for azure functions

`var results = await _schemaProvider.ExecuteRequestAsync(query, _dbContext, HttpContext.RequestServices, null);`

i see that theres a separate method `ExecuteRequestWithContextAsync`, the documentation is not up to date, but thats fine.
the main issue is that i dont have IServiceProvider. I also dont see a reason for this library to require `authService` or `webHostEnvironment ` when the only purpose of this method should be to fetch graphql data from db.

i played around with asp.net implementation, works perfectly, i wish i could add it as another function next to my other functions.

**Comment by bzbetty** [comment]
There's an old example in the repo here - https://github.com/EntityGraphQL/EntityGraphQL/tree/master/src/examples/AzureFunctionApp

It's fairly easy to update to dotnet-isolated if needed though.

**Comment by taipignas** [comment]
it worked! thanks! i swear i tried the same thing, but i guess i missed some small part. thanks a lot for quick help!

**Comment by taipignas** [comment]
i have an additional small question, not sure if i should create a separate issue.
would it be possible to enable paging, filtering and sorting for all tables on schema level? is it an additional feature request, maybe its planned already or the decision is already made not to have it?

**Comment by lukemurray** [comment]
Today, you would need to add
```
[UseFilter]
[UseSort]
[UseConnectionPaging]
// or, use only 1 [UseOffsetPaging]
```

on all the list fields you want it on in your DbContext.

Feel free to add a request to have the an option in `SchemaBuilder`

**Comment by bzbetty** [comment]
SchemaBuilder has OnFieldCreated event that might allow you to do it? Maybe it's new?

```
   // Add Sort to all list fields
   OnFieldCreated = (field) =>
    {
        if (field.ReturnType.IsList && field.ReturnType.SchemaType.GqlType == GqlTypes.QueryObject && !field.FromType.IsInterface)
        {
            field.UseSort();
        }
    }
```

**Comment by lukemurray** [comment]
Ahh yes, that is even in the docs :) That is another way too. Thanks

**Comment by taipignas** [comment]
it is not that straightforward, i need more help if you dont mind.

should filter and paging work with same configuration?

```
if (field.ReturnType.IsList && field.ReturnType.SchemaType.GqlType == GqlTypes.QueryObject && !field.FromType.IsInterface)
{
    field.UseFilter();
    field.UseSort();
    field.UseOffsetPaging();
}
```

UseConnectionPaging() doesnt work for me

```
{
    "errors": [
        {
            "message": "Field 'documents' - No generic method 'Select' on type 'System.Linq.Enumerable' is compatible with the supplied type arguments and arguments. No type arguments should be provided if the method is non-generic. "
        }
    ]
}
```

UseOffsetPaging() partly works but for child properties it is very very slow. much slower than without paging.
GraphQL no longer uses JOIN operation

```
{
  documents(take: 5, skip: 5){
    items{
        id
        attachments{
            totalItems
        }
    }
  }
}
```


this query fetches all attachment data

SELECT [a].[Id], [a].[AttachmentType], [a].[CreatedAt],....
FROM [Attachments] AS [a]

and if i try to add any items under attachments i get an error:
"Field 'documents' - Could not find arguments for field attachments in compile context."

also is the schema provider different for ASP.NET DI? for example in documentation i see method 'ReplaceField' but its not evailable on my schema. Also sometimes variable is called schemaProvider sometimes schema, but the class is (same?) SchemaProvider.

I tried looking through tests, but sadly i cannot install ANTLR on my machine.

**Comment by taipignas** [comment]
i got it to work with LEFT JOIN, but the configuration is not for all tables/fields:
```
schema.Query().ReplaceField("documents",
      db => db.Documents.OrderBy(e => e.Id),
      "Get a page of movies"
    )
    .UseConnectionPaging();
```

it would be nice to have one method on options or schema level: like schemaBuilderOptions.UseOffsetPaging() and it would properly use paging everywhere where possible.

**Comment by lukemurray** [comment]
Sorry, reading it again, there is more

>and if i try to add any items under attachments i get an error:
"Field 'documents' - Could not find arguments for field attachments in compile context."

This is a bug as well, it should produce a valid expression

>also is the schema provider different for ASP.NET DI? for example in documentation i see method 'ReplaceField' but its not evailable on my schema. Also sometimes variable is called schemaProvider sometimes schema, but the class is (same?) SchemaProvider.

`SchemaProvider` is the same. `ReplaceField` is on a `ISchemaType` or `SchemaType<T>`. `SchemaProvider` is the full schema. root fields exist on the schema type `Query` so you can `schema.Query().ReplaceField()`.

`SchemaProvider` provides 3 helpers to 3 main types defined in the GraphQL spec `.Query()`, `.Mutation()` and `.Subscription()`. You can access all other types like `schema.Type<Document>().ReplaceField()`

**Comment by taipignas** [comment]
i have indexes, foreign keys setup properly. i think the problem is when the structure is 3 layers deep, then with paging it adds either items or edges + node to each layer and then it cannot translate properly back to original structure.
if i explicitly setup paging with ReplaceField i do it for top layer only and that works.
So to reproduce it you could create 3 layer deep structure and use recursive paging setup. I get different errors with both UseOffsetPaging and UseConnectionPaging
```
if (field.ReturnType.IsList && field.ReturnType.SchemaType.GqlType == GqlTypes.QueryObject && !field.FromType.IsInterface)
{
    field.UseFilter();
    field.UseSort();
    field.UseOffsetPaging();
}
```

**Comment by lukemurray** [comment]
Ahh, right, yes that is adding paging to _every_ field. Which you may not want.

If you only want it on the root fields (fields on the root GraphQL query type) you could check the `FromType`

```cs
if (field.FromType.TypeDotnet == field.Schema.QueryContextType && field.ReturnType.IsList && field.ReturnType.SchemaType.GqlType == GqlTypes.QueryObject && !field.FromType.IsInterface)
{
    field.UseFilter();
    field.UseSort();
    field.UseOffsetPaging();
}
```

A helper method for knowing if the field is a root field might help. But that should work today.

**Comment by taipignas** [comment]
Thanks a lot for the help, i appreciate it. I want all the functionality to work, i would probably use filter+sort+paging on all properties if available and if the queries were well optimized. But yeah paging on top level is most important. I got started with GraphQL recently on a completely different architecture project and i would like to do the same with .net. There we use postgraphile, it has its own drawbacks but layered queries with all settings seems to work better. It also has a nice feature 'queryCost', it calculates query's cost and its also possible to set a limit on it.

**Comment by bzbetty** [comment]
The query should already be fairly optimized based on how the library works, it would be useful if you could do some sql profiling to show what it is executing or create a small reproduction of the issue in a sample project.

**Comment by taipignas** [comment]
i created a project for reproduction: https://github.com/taipignas/EntityGraphQLLayers

couple queries to try out
```
{
   projects{
    totalItems
    items{
        id
        documents{
            totalItems
        }
    }
   }
}
```

```
{
   projects{
    totalItems
    items{
        id
        documents{
            items{
                id
            }
        }
    }
   }
}

```

**Comment by bzbetty** [comment]
ok, so if I disable the pagination and run this query

```
{
   projects{
	id
        documents{
            id
        }
   }
}
```

then i get 3 queries

```
SELECT CASE
    WHEN EXISTS (
        SELECT 1
        FROM [Projects] AS [p]) THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END

SELECT [p].[Id], [d].[Id]
FROM [Projects] AS [p]
LEFT JOIN [Documents] AS [d] ON [p].[Id] = [d].[ProjectId]
ORDER BY [p].[Id]

SELECT CASE
    WHEN EXISTS (
        SELECT 1
        FROM [Projects] AS [p]) THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END
```

not sure what the select case is for, but can see the nice left join in the middle query.

enabling pagination and running the following modified query (need to do it with an equiv once the first bug is fixed)

```
{
   projects{
    totalItems
    items{
        id
        documents{
            totalItems
        }
    }
   }
}
```

I get 5 queries

```
SELECT CASE
    WHEN EXISTS (
        SELECT 1
        FROM [Projects] AS [p]) THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END

SELECT COUNT(*)
FROM [Projects] AS [p]

SELECT [d].[Id], [d].[DocumentType], [d].[ProjectId]
FROM [Documents] AS [d]

SELECT [p].[Id]
FROM [Projects] AS [p]

SELECT CASE
    WHEN EXISTS (
        SELECT 1
        FROM [Projects] AS [p]) THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END
```

2 of those select case statements
1 to grab the total count
then separate ones for Documents and Projects (not joined) and weirdly pulling back all projects even though we only wanted the count.

Yeah there's definitely a bug here, should probably be split off into a new issue.

**Comment by bzbetty** [comment]
oh interesting, it's because project.documents is an ICollection and therefore not treated as a IQueryable  (changing the property to IQueryable makes it a SQL count - although no idea what else that then breaks)

(eg it's using the IEnumerable Count instead of the IQueryable Count and therefore pulls back the full table).

That's rather interesting as
* It's the recommended type to use efcore (at least the documentation always uses ICollection)
* It's smart enough to merge them together into one query against the documents table, but not smart enough to realise it's IQueryable

**Comment by lukemurray** [comment]
Interesting. Thanks @bzbetty and @taipignas.

The expression also look wrong as it is using the db context for documents? `p_DocumentDbContext.Documents.Count()` should be `p_Project.Documents.Count()`.

It seems the way the paging expressions work, is it builds expressions _assuming_ they are a top level property.

**Comment by bzbetty** [comment]
oh nice catch! yeah looking at the json response it's totally putting the total counts as the table count and not the related files

**Comment by lukemurray** [comment]
@taipignas I couldn't get that project you shared running, something about `The gRPC channel URI 'http://:' could not be parsed.`. I haven't dug into it.

I think it may be related to and fixed by the changes for #358 where it could reuse other paging expressions. I added a test but that passes and it passed with 5.2 as well.

Any help reproducing this would be great, or confirming that the fix in master does resolve your issues. Thanks.

**Comment by lukemurray** [comment]
I'm closing this as I cannot currently reproduce. I believe #358 fixed the issue with the reuse of the expression. Please open a new issue if you find a reproduction.
